File: ts/11.binary-tree/find-closest-value.test.ts
import { assert, describe, test } from 'vitest';
import { BinaryTreeImpl } from './BinaryTree';
import { NodeImpl } from './NodeImpl';
import { Node } from './Node';
import { findClosestNumber } from './find-closest-value';

describe('find closest value', () => {
    test('first test case', () => {
        let binaryTree = new BinaryTreeImpl<number>(new NodeImpl(10));
        let root = binaryTree.getRoot() as Node<number>;
        let right = root.pushRightLeaf(15);
        let left = root.pushLeftLeaf(5);
        left.pushRightLeaf(5);
        left.pushLeftLeaf(2).pushLeftLeaf(1);
        right.pushRightLeaf(22);
        right.pushLeftLeaf(13).pushRightLeaf(14);

        let result = findClosestNumber(binaryTree, 12);

        assert.equal(result, 13);
    });
});

File: golang/11.binary-tree/main/find_closest_value_test.go
package main

import (
	"github.com/AmirAhmadzadeh/binary_tree/binary_tree"
	"testing"
)

func TestClosestValue(t *testing.T) {
	binaryTree := binary_tree.New(10)
	root := binaryTree.GetRoot()

	right := root.PushRightLeaf(15)
	left := root.PushLeftLeaf(5)
	left.PushRightLeaf(5)
	left.PushLeftLeaf(2).PushLeftLeaf(1)
	right.PushRightLeaf(22)
	right.PushLeftLeaf(13).PushRightLeaf(14)

	result := findClosestValueInBinaryTree(binaryTree, 12)

	if result != float64(13) {
		t.Errorf("** first test case ** Expected (%v) is not same as"+
			" actual (%v)", 13, result)
	}
}

func TestClosestValueDfs(t *testing.T) {
	binaryTree := binary_tree.New(10)
	root := binaryTree.GetRoot()

	right := root.PushRightLeaf(15)
	left := root.PushLeftLeaf(5)
	left.PushRightLeaf(5)
	left.PushLeftLeaf(2).PushLeftLeaf(1)
	right.PushRightLeaf(22)
	right.PushLeftLeaf(13).PushRightLeaf(14)

	result := findClosestValueInBinaryTreeDfs(binaryTree, 12)

	if result != float64(13) {
		t.Errorf("** dfs version ** Expected (%v) is not same as"+
			"(%v)", 13, result)
	}
}
